### PR TITLE
feat(ExTagLink): add fanbox links

### PR DESCRIPTION
**Pull request by korafu**
this PR should add support for adding pixiv fanbox links to the tag area on Ex. let me know if this change needs revising or if it is unwanted!

### CODE REVIEW

File: userscripts/ExTagLink.user.js [hunk 1]
```diff
@@ -74,10 +75,27 @@ for (const com of $$('.c6')) {
...
+      fanboxLinks.push(fanboxLink);
+    }
+
     textNode = walker.nextNode();
   }
 }
 
+if (fanboxLinks.length) {
+  const linkItems = fanboxLinks.map((link) => ({
+    name: new URL(link).hostname.split('.')[0],
```
Comment (FlandreDaisuki) [inline userscripts/ExTagLink.user.js]: I found there are 2 patterns in comments:

- `https://www.fanbox.cc/@lambda`
- `https://lambda.fanbox.cc/`

I think we can show the name `lambda` for both patterns

File: userscripts/ExTagLink.user.js [hunk 2]
```diff
@@ -74,10 +75,27 @@ for (const com of $$('.c6')) {
       pixivLinks.push(pixivLink);
     }
 
+    const fanboxLink = textNode.textContent.match(
+      /https:\/\/\S+\.fanbox\.cc\/\S+/g
```
Comment (FlandreDaisuki) [inline userscripts/ExTagLink.user.js]: This won't match `https://lambda.fanbox.cc/` because of the last `\S+`.

**Comment by FlandreDaisuki** [comment]
Great idea! Please let me know if you need any help.

File: userscripts/ExTagLink.user.js (hunk 1 above)
Comment (korafu) [reply userscripts/ExTagLink.user.js]: good call, I was not aware of the other URL pattern. I hope I covered it with the latest update!

**Comment by korafu** [comment]
thanks for the review! I updated the regex to work on fanbox URLs with the following patterns:

```
https://lambda.fanbox.cc/
https://lambda.fanbox.cc/posts/123
https://fanbox.cc/@lambda
https://fanbox.cc/@lambda/posts/123
```

on a side note, I saw ESLint and your ESLint rules package in package.json, so installed it but I see a bunch of lint errors now:

image

if you would like, I can run the linter in a separate PR for you to review (and maybe set up auto linting as well?)

**Comment by FlandreDaisuki** [comment]
I found one more other pattern for fanbox.

```diff
  https://lambda.fanbox.cc/
  https://lambda.fanbox.cc/posts/123
  https://fanbox.cc/@lambda
  https://fanbox.cc/@lambda/posts/123
+ https://www.fanbox.cc/@lambda
```

so, I think the following regex can parse first:
```js
/https?:\/\/([\w-]+\.)?fanbox.cc\/(?:@([\w-]+))?/mg
//          ^~ 1st group              ^~ 2nd group
```
https://regex101.com/r/kVpNwE/1

then, if we have second group, use second group. Else, use first group.

---

I'll fix the lint rule in future. You can ignore it currently 😓

**Comment by korafu** [comment]
I believe the original regex already accounted for optional `www.` in the URL, and I have updated it to fix a bug I had in the post detection

https://regex101.com/r/Acn6Px/1

now the regex match array will include the entire URL with `/posts/<post ID>` so the logic on L99-114 should work correctly now!

**Comment by FlandreDaisuki** [comment]
Ok, I think there is a misunderstanding.

What I expected was to show the artist name from the following links, but I only see 'www'.

https://exhentai.org/g/2757520/fc42e712c9/
https://exhentai.org/g/2757482/df77cc8161/
https://exhentai.org/g/2757469/2d57af1548/
https://exhentai.org/g/2757413/70f2d16235/

That is why I provided the regex above.

File: userscripts/ExTagLink.user.js [hunk 3]
```diff
@@ -68,16 +69,53 @@ for (const com of $$('.c6')) {
...
+      // match https://fanbox.cc/@lambda/posts/123
+      name = link.match(/posts\/(\d+)/)[1];
+    } else if (link.includes('fanbox.cc/@')) {
+      // match https://fanbox.cc/@lambda
+      name = link.match(/@(\S+)/)[1];
+    } else {
+      // match https://lambda.fanbox.cc/
+      name = new URL(link).hostname.match(/([\w-]+)\.fanbox\.cc/)[1];
+    }
+    return { name, link };
```
Comment (FlandreDaisuki) [inline userscripts/ExTagLink.user.js]: Ok, I think there is a misunderstanding. I'm not arguing about your regex on matching fanbox link.

What I expected was to show the artist name from all the url patterns, like following links, but I only see `www`.

https://exhentai.org/g/2757520/fc42e712c9/
https://exhentai.org/g/2757482/df77cc8161/
https://exhentai.org/g/2757469/2d57af1548/
https://exhentai.org/g/2757413/70f2d16235/

That is why I provided the regex above.

Do you expect showing `www` as name?

Comment (FlandreDaisuki) [reply userscripts/ExTagLink.user.js]: I think the following formats are more intuitive:

For links like:

- `https://lambda.fanbox.cc/posts/123`
- `https://fanbox.cc/@lambda/posts/123`

The name could be `lambda:123` or `lambda/123` or `@lambda:123`

For links like:  

- `https://lambda.fanbox.cc/`
- `https://fanbox.cc/@lambda`
- `https://www.fanbox.cc/@lambda`

The name would be `lambda` or `@lambda`

**Comment by korafu** [comment]
my mistake, I understand now! I believe I have updated it properly this time

**Comment by FlandreDaisuki** [comment]
I test several galleries and look good to me. :+1:

**Pull request merged**